projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/vector/LLVMFunctionVector.java

```diff
@@ -0,0 +1,128 @@
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.runtime.vector;
+
+import java.util.Arrays;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.llvm.runtime.LLVMAddress;
+import com.oracle.truffle.llvm.runtime.LLVMFunctionDescriptor;
+
+public final class LLVMFunctionVector {
+    private final Object[] vector;
+
+    public static LLVMFunctionVector create(LLVMFunctionDescriptor[] vector) {
+        return new LLVMFunctionVector(vector);
+    }
+
+    public static LLVMFunctionVector create(long[] vector) {
+        return new LLVMFunctionVector(vector);
+    }
+
+    public static LLVMFunctionVector create(Object[] vector) {
+        return new LLVMFunctionVector(vector);
+    }
+
+    public static LLVMFunctionVector createNullVector() {
+        return new LLVMFunctionVector();
+    }
+
+    private LLVMFunctionVector(LLVMFunctionDescriptor[] vector) {
+        this.vector = new Object[vector.length];
+        for (int i = 0; i < vector.length; i++) {
+            this.vector[i] = vector[i];
+        }
+
+    }
+
+    private LLVMFunctionVector(long[] vector) {
+        this.vector = new LLVMAddress[vector.length];
+        for (int i = 0; i < vector.length; i++) {
+            this.vector[i] = LLVMAddress.fromLong(vector[i]);
+        }
+    }
+
+    private LLVMFunctionVector(Object[] vector) {
+        this.vector = new Object[vector.length];
+        for (int i = 0; i < vector.length; i++) {
+            this.vector[i] = vector[i];
+        }
+    }
+
+    private LLVMFunctionVector() {
+        this.vector = null;
+    }
+
+    public Object[] getValues() {
+        return vector;
+    }
+
+    public LLVMAddress[] getAddresses() {
+        LLVMAddress[] addresses = new LLVMAddress[vector.length];
+        for (int i = 0; i < vector.length; i++) {
+            addresses[i] = getAddress(i);
+        }
+        return addresses;
+    }
+
+    public Object getValue(int index) {
+        return vector[index];
+    }
+
+    public LLVMAddress getAddress(int index) {
+        if (vector[index] instanceof LLVMAddress) {
+            return (LLVMAddress) vector[index];
+        } else if (vector[index] instanceof LLVMFunctionDescriptor) {
+            LLVMFunctionDescriptor fd = (LLVMFunctionDescriptor) vector[index];
+            LLVMAddress ptr = LLVMAddress.fromLong(fd.toNative().asPointer());
```

Comment:
Please don't introduce new users of `toNative()`, especially on the fast-path. We're currently trying to get rid of usages of this function. See https://github.com/graalvm/sulong/blob/master/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/LLVMFunctionDescriptor.java#L479

The proper way to do this on the fast-path would be to use a `LLVMToNativeNode`. Then you also wouldn't need the `if` cascade, the `LLVMToNativeNode` would automatically specialize itself to the correct type.